### PR TITLE
Feature/SpiralIterator

### PR DESCRIPTION
**Pull request by karlthepagan**
The TS version of this is pretty well tested. 

The dirTransform step possibly could be replaced using the method that Traveler.js does (arrays)

### CODE REVIEW

**Comment by karlthepagan** [comment]
@chobobobo  and I were discussing find vs lookForAt and the engine code definitely makes it seem that lookForAt will save us a lot for driveby repairs.

This tool will let you iterate in a circular spiral out from the center.

File: util.xyIterator.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+ *  19 P
+ *  18 O  9  A  B  C
+ *  17 N  8  1  2  D
+ *  16 M  7  0  3  E
+ *  15 L  6  5  4  F
+ *  14 K  J  I  H  G
+ *
+ *  Easy starting-points are 0, 1, A, etc. Use the ring number to indicate these.
+ */
+class XYIterator {
```
Comment (Spedwards) [inline util.xyIterator.js]: Load this into `Util`, same as the DiamondIterator

File: util.xyIterator.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        this._dir = TOP;
+        this.x = xy.x;
+        if (ring === undefined) {
+            this._ring = 0;
+            this._step = 2;
+            this.y = xy.y + 1;
+        } else {
+            this._ring = ring - 1;
+            this._step = 1;
+            this.y = xy.y;
```
Comment (Spedwards) [inline util.xyIterator.js]: For simplicity's sake, move `this.y = xy.y` before the if-statement, then inside `ring === undefined`, increment the value (`this.y++`)

File: util.xyIterator.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        return {
+            done: false,
+            value: {x: this.x, y: this.y},
+        };
+    }
+
+    static dirTransform(origin, dir)
+    {
```
Comment (Spedwards) [inline util.xyIterator.js]: Should not be on it's own line.

File: util.xyIterator.js (hunk 2 above)
Comment (karlthepagan) [inline util.xyIterator.js]: no demonstrable value

many devs have picked up [SSA habits](https://en.wikipedia.org/wiki/Static_single_assignment_form) and will argue to death the opposite

File: util.xyIterator.js (hunk 1 above)
Comment (karlthepagan) [inline util.xyIterator.js]: I'm also going to rename this to spiral.iterator

**Comment by karlthepagan** [comment]
The only other thing I could wish for this is to add variations which let you advance thru different directions, but that's more effort than its worth.

**Comment by chobobobo** [comment]
How does this know when to stop?

**Comment by karlthepagan** [comment]
This one never stops. That's probably an important API feature.

**Pull request merged**